### PR TITLE
feat: add webSocketURL option

### PR DESCRIPTION
**Pull request by anshumanv**
- [x] This is a **bugfix**
- [ ] This is a **feature**
- [x] This is a **code refactor**
- [x] This is a **test update**
- [x] This is a **docs update**
- [x] This is a **metadata update**

### For Bugs and Features; did you add new tests?
migrated existing

### Motivation / Use-Case
Move out client.host/port/path to a new option `webSocketURL`

### Breaking Changes
Yes

### CODE REVIEW

File: lib/options.json [hunk 1]
```diff
@@ -285,6 +263,40 @@
...
+              "anyOf": [
+                {
+                  "type": "number"
+                },
+                {
+                  "type": "string"
+                },
+                {
+                  "enum": ["auto"]
+                }
```
Comment (snitin315) [inline lib/options.json:292]:
```suggestion
```

File: lib/utils/DevServerPlugin.js [hunk 2]
```diff
@@ -35,34 +35,46 @@ class DevServerPlugin {
...
-    }
+    let webSocketURL = '';
+
+    if (typeof options.webSocketURL === 'string') {
+      webSocketURL = options.webSocketURL;
+    } else if (typeof options.webSocketURL === 'object') {
+      /** @type {string} */
+      let host = '';
+
+      if (options.webSocketURL && options.webSocketURL.host) {
```
Comment (snitin315) [inline lib/utils/DevServerPlugin.js:46]: Already check above for object

```suggestion
      if (options.webSocketURL.host) {
```

File: lib/utils/DevServerPlugin.js [hunk 3]
```diff
@@ -35,34 +35,46 @@ class DevServerPlugin {
...
+      let port = '';
 
-    if (options.client && options.client.port) {
-      port = `&port=${options.client.port}`;
-    } else if (options.webSocketServer.options.port && !isSockJSType) {
-      port = `&port=${options.webSocketServer.options.port}`;
-    } else {
-      port = `&port=${options.port}`;
-    }
+      if (options.webSocketURL && options.webSocketURL.port) {
```
Comment (snitin315) [inline lib/utils/DevServerPlugin.js:55]:
```suggestion
      if (options.webSocketURL.port) {
```

File: lib/utils/DevServerPlugin.js [hunk 4]
```diff
@@ -35,34 +35,46 @@ class DevServerPlugin {
...
-    /** @type {string} */
-    let path = '';
+      /** @type {string} */
+      let path = '';
+
+      // Only add the path if it is not default
+      if (
+        options.webSocketURL &&
+        options.webSocketURL.path &&
+        options.webSocketURL.path
```
Comment (snitin315) [inline lib/utils/DevServerPlugin.js:70]:
```suggestion
        options.webSocketURL.path
```

File: lib/utils/DevServerPlugin.js [hunk 5]
```diff
@@ -35,34 +35,46 @@ class DevServerPlugin {
...
+      ) {
+        path = `&path=${options.client.path}`;
```
Comment (snitin315) [inline lib/utils/DevServerPlugin.js:72]:
```suggestion
        path = `&path=${options.webSocketURL.path}`;
```

**Comment by anshumanv** [comment]
no need review drafts, let me finish

File: lib/options.json (hunk 1 above)
Comment (anshumanv) [reply lib/options.json:292]: why?

Comment (snitin315) [reply lib/options.json:292]: As per the thread in slack. 

> webSocketURL: STRING | { host: string, port: number, path: string }

File: lib/options.json [hunk 6]
```diff
@@ -306,6 +284,34 @@
       "description": "Specifies client properties. https://webpack.js.org/configuration/dev-server/#devserverclient",
       "additionalProperties": false
     },
+    "webSocketURL": {
```
Comment (alexander-akait) [inline lib/options.json:287]: It should be in the `client` options, because this option only for client

File: lib/utils/DevServerPlugin.js [hunk 7]
```diff
@@ -35,36 +35,44 @@ class DevServerPlugin {
     // TODO show warning about this
     const isSockJSType = options.webSocketServer.type === 'sockjs';
 
-    /** @type {string} */
-    let host = '';
+    let webSocketURL = '';
+
+    if (typeof options.webSocketURL === 'string') {
+      webSocketURL = options.webSocketURL;
```
Comment (alexander-akait) [inline lib/utils/DevServerPlugin.js:41]: We need normalize `webSocketURL`, i.e. parse and get `host`/`port`/`path`

Comment (alexander-akait) [reply lib/utils/DevServerPlugin.js:41]: Same for below code

File: test/server/__snapshots__/Server.test.js.snap.webpack5 [hunk 8]
```diff
@@ -45,8 +43,7 @@ Array [
   Array [
     "client",
     "index.js?http:",
-    "0.0.0.0&path=",
-    "ws&port=8100",
+    "0.0.0.0",
   ],
   Array [
```
Comment (alexander-akait) [inline test/server/__snapshots__/Server.test.js.snap.webpack5:48]: Existing snapshots should not be changed, it is big breaking change

Comment (anshumanv) [reply test/server/__snapshots__/Server.test.js.snap.webpack5:48]: oh yep need normalize

File: lib/utils/DevServerPlugin.js (hunk 7 above)
Comment (anshumanv) [reply lib/utils/DevServerPlugin.js]: fixed

File: lib/options.json (hunk 6 above)
Comment (anshumanv) [reply lib/options.json]: yep maybe I was confused before, fixed now 👍

**Comment by anshumanv** [comment]
@alexander-akait @snitin315 CI should be green now 😎

File: lib/options.json [hunk 9]
```diff
@@ -301,6 +279,34 @@
...
           "description": "Tells devServer to inject a Hot Module Replacement entry."
+        },
+        "webSocketURL": {
+          "anyOf": [
+            {
+              "type": "string"
+            },
+            {
+              "type": "object",
+              "properties": {
```
Comment (snitin315) [inline lib/options.json:290]:
```suggestion
              "type": "object",
              "additionalProperties": false,
              "properties": {
```

File: lib/options.json [hunk 10]
```diff
@@ -301,6 +279,34 @@
...
+                  "type": "string",
+                  "description": "Tells clients connected to devServer to use the provided path to connect."
+                },
+                "port": {
+                  "anyOf": [
+                    {
+                      "type": "number"
+                    }
+                  ],
+                  "description": "Tells clients connected to devServer to use the provided path to connect."
```
Comment (snitin315) [inline lib/options.json:305]:
```suggestion
                  "description": "Tells clients connected to devServer to use the provided port."
```

File: lib/utils/DevServerPlugin.js [hunk 11]
```diff
@@ -35,36 +35,44 @@ class DevServerPlugin {
     // TODO show warning about this
     const isSockJSType = options.webSocketServer.type === 'sockjs';
 
-    /** @type {string} */
-    let host = '';
+    let webSocketURL = '';
+
+    if (typeof options.client.webSocketURL === 'string') {
+      webSocketURL = options.client.webSocketURL;
```
Comment (snitin315) [inline lib/utils/DevServerPlugin.js:41]: So user input should be like this in case of string `&host=0.0.0.0&port=8080&path=foo`?

Comment (anshumanv) [reply lib/utils/DevServerPlugin.js:41]: Yeah I think that's the only possible way since it's resolving a client path before it, not very convenient for user I agree

/cc @alexander-akait should confirm

File: lib/options.json (hunk 10 above)
Comment (anshumanv) [reply lib/options.json:305]: yep lets improve it

File: lib/utils/DevServerPlugin.js (hunk 11 above)
Comment (anshumanv) [reply lib/utils/DevServerPlugin.js:41]: It doesn't make sense actually, most likely user will pass an absolute URL and we can use it directly

Comment (snitin315) [reply lib/utils/DevServerPlugin.js:41]: It should work like `public`. And then we remove options.public in favor of `webSocketURL`

Comment (alexander-akait) [reply lib/utils/DevServerPlugin.js:41]: > It should work like public. And then we remove options.public in favor of webSocketURL

Yes, right direction

Comment (anshumanv) [reply lib/utils/DevServerPlugin.js:41]: yep lets update

**Comment by alexander-akait** [comment]
@anshumanv Do you need help? I can finish, it is blocker for the next `beta` release

**Comment by anshumanv** [comment]
@alexander-akait should be green now, feel free to make any changes or lmk I can do it, thanks for help 👍

**Review by snitin315 — COMMENTED** [review]
Need to remove `public` option.

File: test/e2e/ClientOptions.test.js [hunk 12]
```diff
@@ -38,7 +38,7 @@ for (const webSocketServerType of webSocketServerTypes) {
     });
   };
 
-  describe(`should work behind proxy, when hostnames are same and ports are different (${webSocketServerType})`, () => {
+  describe.only(`should work behind proxy, when hostnames are same and ports are different (${webSocketServerType})`, () => {
```
Comment (bompus) [inline test/e2e/ClientOptions.test.js:41]: don't forget to remove the .only :)

Comment (anshumanv) [reply test/e2e/ClientOptions.test.js]: oh yep, good catch!

**Comment by alexander-akait** [comment]
Let's finish :+1:

**Comment by alexander-akait** [comment]
Finished... We still have some edge cases and places where we can simplify code, but I will do it in the next PRs

**Comment by snitin315** [comment]
Also need to remove `--public` from cli-flags.

**Comment by alexander-akait** [comment]
Oh, yes, forgot about it

**Comment by alexander-akait** [comment]
I will remove this option in this PR and let's add all options here https://github.com/webpack/webpack-dev-server/pull/3325

**Comment by alexander-akait** [comment]
All our tests are complete horror, we need to rewrite them, I spent half a day to finish them

**Comment by alexander-akait** [comment]
@snitin315 @anshumanv Can you review?

File: test/server/utils/normalizeOptions.test.js [hunk 13]
```diff
@@ -85,7 +87,9 @@ describe('normalizeOptions', () => {
       multiCompiler: false,
       options: {
         client: {
-          path: '/custom/path/',
+          webSocketURL: {
+            path: '/custom/path/',
+          },
```
Comment (snitin315) [inline test/server/utils/normalizeOptions.test.js:92]: We should add one case with `webSocketURL: 'ws://myhost.com:8080/path'`, and check it normalizes host, port and path correctly from the string.

File: lib/utils/normalizeOptions.js [hunk 14]
```diff
@@ -104,6 +104,18 @@ function normalizeOptions(compiler, options, logger) {
...
   }
 
+  if (typeof options.client.webSocketURL === 'undefined') {
+    options.client.webSocketURL = {};
+  } else if (typeof options.client.webSocketURL === 'string') {
+    const parsedURL = new URL(options.client.webSocketURL);
+
+    options.client.webSocketURL = {
+      host: parsedURL.hostname,
+      port: parsedURL.port,
```
Comment (snitin315) [inline lib/utils/normalizeOptions.js:114]: Hmm, I think port here will be type `string`,  schema have only type `number`

File: test/server/utils/normalizeOptions.test.js (hunk 13 above)
Comment (bompus) [reply test/server/utils/normalizeOptions.test.js:92]: Will the webSocketURL string allow me to provide wss:// and actually have the browser request using wss://, or will the normalization turn it back into ws:// if https is false? I'm running a reverse https -> http proxy in front, so the dev server should be https:false, but currently doing so, the websocket logic attempts to use ws:// even though the address bar is an https:// url. Just checking to see if this string webSocketURL option will indeed allow me to now force wss:// to be used for the websocket requests.

Comment (alexander-akait) [reply test/server/utils/normalizeOptions.test.js:92]: @bompus you can use `ws` and `wss`, and even `http`/`https` (we normalize them to `ws`/`wss` for backward compatibility). But you should use `webSocketURL` only when you proxy dev server, if you set `https: true` we will use `wss` by default

Comment (alexander-akait) [reply test/server/utils/normalizeOptions.test.js:92]: @snitin315 we have this test (in `test/e2e/web-socket-server-and-url.test.js`):
```
client: {
  webSocketURL: `ws://myhost.test:${port2}/foo/test/bar/`,
},
`

File: lib/utils/normalizeOptions.js (hunk 14 above)
Comment (alexander-akait) [reply lib/utils/normalizeOptions.js:114]: Yes, I think about it, here two solution:
- allow to set port as `string` everywhere
- do not allow use `port` as `string` everywhere

Let's fix it in other PR, I found we allow to set `host` to `null`, it is wrong, also we need allow to set port as `string` for websocket server too

File: test/server/utils/normalizeOptions.test.js (hunk 13 above)
Comment (snitin315) [reply test/server/utils/normalizeOptions.test.js:92]: All good then 👍🏻

File: lib/utils/normalizeOptions.js (hunk 14 above)
Comment (snitin315) [reply lib/utils/normalizeOptions.js:114]: > allow to set port as string everywhere

+1, I can send a PR for the same 👍🏻

**Review by snitin315 — APPROVED** [review]

**Review by anshumanv — COMMENTED** [review]
looks good 👍

File: test/server/utils/normalizeOptions.test.js (hunk 13 above)
Comment (bompus) [reply test/server/utils/normalizeOptions.test.js:92]: @alexander-akait I tried this on master after it was merged and browser is still told to request ws:// instead of wss:// , even when wss:// is specified on webSocketURL. Am I missing something in the config, or is this as expected?

My config is:
```
devServer: {
      https: false,
      host: '127.0.0.1',
      port: 8888,
      firewall: false,
      client: {
        webSocketURL: 'wss://myhost.test/ws/'
      }
}
```

Comment (alexander-akait) [reply test/server/utils/normalizeOptions.test.js:92]: hm, why you need to change `protocol`?

Comment (alexander-akait) [reply test/server/utils/normalizeOptions.test.js:92]: webpack dev server without https and proxy server under https?

Comment (bompus) [reply test/server/utils/normalizeOptions.test.js:92]: The dev server is proxied via a protected cloudflared argo tunnel (cloudflare https -> nginx http -> webpack-dev-server http), so the local dev server needs to listen for http instead of https. This is done so dev matches production closer to prevent surprises.

This works in a previous version of v4 (pre-beta3?) and also in v3.

It would be useful if you just detected the window.location.protocol and used that as the default for ws:// or wss://, which is probably what was done previously. With the recent changes, it seems that everything is being toggled off that https:[true/false] option.

I'm running https:true for now to resolve the issue, but it'd be nice to go back to the old config so the reverse proxy matches production and hits all the same cloudflare https -> nginx http logic.

Comment (alexander-akait) [reply test/server/utils/normalizeOptions.test.js:92]: @bompus fair enough, my idea - adding `protocol` option to `webSocketURL` (you can use it in `string` notation too), also provide `protocol: 'auto'` to use from `location`, so passing `auto://0.0.0.0:0/` will means all from `location`, what do you think?

Comment (bompus) [reply test/server/utils/normalizeOptions.test.js:92]: @alexander-akait that sounds good to me, as I had already expected it to use `ws` or `wss` based on what had been specified in webSocketURL. `auto` works as well, and I assume one would need to use `auto://0.0.0.0:0/wspath/` if they wanted a specific path?

Comment (alexander-akait) [reply test/server/utils/normalizeOptions.test.js:92]: >  I assume one would need to use auto://0.0.0.0:0/wspath/ if they wanted a specific path?

Yes, maybe make sense to add `auto` to path too

Comment (bompus) [reply test/server/utils/normalizeOptions.test.js:92]: Great. This should solve my use case 👍

Comment (alexander-akait) [reply test/server/utils/normalizeOptions.test.js:92]: Put in todo, will solve tomorrow

Comment (bompus) [reply test/server/utils/normalizeOptions.test.js:92]: @alexander-akait did you get a chance to start on this yet, or still pending? Just want to make sure it doesn't get buried.

Comment (alexander-akait) [reply test/server/utils/normalizeOptions.test.js:92]: in my todo on tomorrow, don't worry :+1:
